### ISSUE TITLE
paddle.distributed.fleet.distributed_model 文档问题

### ISSUE THREAD

**Issue by ShaneTian** [PaddlePaddle/Paddle#31370]
- API: `paddle.distributed.fleet.distributed_model`
- 文档版本信息：`2.0`
- 问题：该 API 作用相当于 PyTorch 中的 `torch.nn.parallel.DistributedDataParallel` 还是 `torch.nn.DataParallel`?
- 细节：使用 `fleet` `collective` 模式进行开发时，已使用 `fleet.distributed_optimizer` 对 `optimizer` 和 `fleet.DistributedStrategy` 包了一层。想确认模型是否也需要使用 `fleet.distributed_model` 再包一层？

**Comment by ForFishes** [comment]
您好，这里 `paddle.distributed.fleet.distributed_model` 相当于 `torch.nn.parallel.DistributedDataParallel`。paddle中没有类似于`torch.nn.DataParallel`的接口。
第二个问题，需要将`fleet.distributed_model`在封装一层。原因是动态图主要在`fleet.distributed_model`进行分布式设计，静态图是在 `fleet.distributed_optimizer`进行分布式设计。所以使用上，两个接口都是需要的（如果不区分动态图和静态图）。

**Comment by ShaneTian** [comment]
感谢回复。
相关动态图分布式训练的开发代码主要简化为以下，还麻烦你看看这样是否正确了：
```py
# init
role = fleet.PaddleCloudRoleMaker(is_collective=True)
fleet.init(role)

dist_strategy = fleet.DistributedStrategy()
# some settings about DistributedStrategy ...
optimizer = fleet.distributed_optimizer(optimizer, strategy=dist_strategy)
model = fleet.distributed_model(model)

# some settings about dataset
batch_sampler = DistributedBatchSampler(dataset, batch_size, shuffle=True)
dataloader = DataLoader(dataset, batch_sampler=batch_sampler, num_works=4)
```

**Comment by ForFishes** [comment]
您好，你这个写法是正确的。

**Comment by ShaneTian** [comment]
好的，多谢！

**Comment by ShaneTian** [comment]
抱歉，再多问两个问题：
1. 单机多卡使用上述 `fleet` 代码进行训练时需要只使用 0 号进程保存模型吗？比如当 `dist.get_rank() == 0` 时才 `save`，其他进程不 `save`。如果都 `save` 了的的话发现好像也是只有一份文件，是怎么回事呢？
2. `train.py` 中隔固定步执行了 `evaluation`，请问在 `evaluation` 中：
    1. `output = model(input)` `loss = loss_fn(output, label)` 得到的 `loss` 是当前进程的吧？
    2. 是不是应该 `paddle.distributed.all_reduce(loss, op=ReduceOp.SUM)` 一下得到所有进程的 `loss` 和再逐元素除以进程数 `dist.get_world_size()` 才是 `eval` 数据集的平均 `loss`？（注：`eval_dataloader` 也过了 `DistributedBatchSampler`）

**Comment by ForFishes** [comment]
您好。1、因为是数据并行，各个进程的模型参数几乎没有区别（只有一点细微的保存精度误差），所以那个进程都可以用来保存。但是如果不选择``dist.get_rank() == 0``,那么每个进程都会往同一个文件夹写文件，相互覆盖，所以最终也是只有一个文件。所以为了避免这个情况，推荐使用``dist.get_rank() == 0 ``才save。
2、（1）是的，loss是当前进程的。原因是eval阶段，没有调用反向（loss.backward），所以各个进程间不会进行通信同步。
      （2）对的。如果要拿到loss的平均，是需要按照你的说的做。 补充一点：因为你这里是loss，取平均是没有问题的。但是，如果是auc，由于各个卡的数据量并不是完全相同（最后一个batch的大小可能不相同），可能取平均会有误差，所以一般eval数据量不大的情况下，选择单机单卡的方式更佳。
感谢您的使用。

**Comment by ShaneTian** [comment]
回答非常清晰，感谢！
